File: Sources/Providers/CursorUsage.swift
import Foundation

/// Parses Cursor's `GET /api/usage-summary`, recorded from a live free account:
///
/// ```json
/// { "billingCycleStart": "2026-08-24T03:32:15.933Z",
///   "billingCycleEnd":   "2026-09-24T03:32:15.933Z",
///   "membershipType": "free", "isUnlimited": false,
///   "individualUsage": {
///     "plan": { "enabled": true, "used": 0, "limit": 0, "remaining": 0,
///               "breakdown": { "included": 0, "bonus": 19, "total": 19 },
///               "autoPercentUsed": 0, "apiPercentUsed": 19,
///               "totalPercentUsed": 9.5 },
///     "onDemand": { "enabled": false, "used": 0, "limit": null } } }
/// ```
///
/// Enterprise / team plans ship a different shape — no `plan` percentages,
/// just a hard `overall` ceiling:
///
/// ```json
/// { "membershipType": "enterprise", "limitType": "team",
///   "individualUsage": {
///     "overall": { "enabled": true, "used": 6907, "limit": 45000, "remaining": 38093 } },
///   "teamUsage": {
///     "onDemand": { "enabled": true, "used": 0, "limit": 1000000 } } }
/// ```
///
/// Cursor meters an **allowance, not a request count**. Plan & Usage has two
/// bars. The ring follows **Auto** (`autoPercentUsed`) — Cursor Models,
/// Grok, Composer. API (`apiPercentUsed`) is a separate bucket.
/// `totalPercentUsed` is a blend of the two and is not a row here.
///
/// The `used`/`limit` pair sits at zero on a free plan even while real usage
/// is happening, because the allowance arrives as `breakdown.bonus` rather
/// than as a dollar limit. Reading those reported 0% for an account that was
/// actually on the Cursor Models bar, which is exactly what this parser
/// used to do. Enterprise is the opposite: there is no percentage field, so
/// `used`/`limit` on `overall` is the reading.
enum CursorUsage {
    static let modelsLabel = "Auto usage"

    /// The window the ring should mean. Cursor Models when that field exists,
    /// never the blended total, never API — and on an enterprise/team plan,
    /// which reports neither, the hard `included` ceiling.
    static func headlineID(in windows: [LimitWindow]) -> String {
        if windows.contains(where: { $0.id == "auto" }) { return "auto" }
        if windows.contains(where: { $0.id == "included" }) { return "included" }
        return "api"
    }

    static func windows(fromJSON json: String) throws -> [LimitWindow] {
        guard let data = json.data(using: .utf8),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { throw UsageProviderError.badResponse(status: 0) }

        let resetsAt = date(root["billingCycleEnd"])
        let usage = root["individualUsage"] as? [String: Any] ?? [:]
        let plan = usage["plan"] as? [String: Any] ?? [:]
        let team = root["teamUsage"] as? [String: Any] ?? [:]

        var windows: [LimitWindow] = []

        // Zero is a reading, not an absence — a fresh month is 0% on this bar.
        if let models = percent(plan["autoPercentUsed"]) {
            windows.append(LimitWindow(id: "auto", label: modelsLabel,
                                       usedFraction: models, resetsAt: resetsAt))
        }
        if let api = percent(plan["apiPercentUsed"]), api > 0 {
            windows.append(LimitWindow(id: "api", label: "API usage",
                                       usedFraction: api, resetsAt: resetsAt))
        }
        if let onDemand = spendWindow(usage["onDemand"], id: "on_demand",
                                      label: "On demand", resetsAt: resetsAt) {
            windows.append(onDemand)
        }

        // Enterprise / team plans omit `plan` entirely and meter a hard
        // `overall` ceiling instead. Keep the window id as `included` so the
        // provider's headlineID still resolves.
        if windows.isEmpty,
           let overall = spendWindow(usage["overall"], id: "included",
                                     label: "Included usage", resetsAt: resetsAt) {
            windows.append(overall)
        }
        if let teamOnDemand = spendWindow(team["onDemand"], id: "team_on_demand",
                                          label: "Team on demand", resetsAt: resetsAt),
           (teamOnDemand.usedFraction ?? 0) > 0 {
            windows.append(teamOnDemand)
        }

        guard windows.isEmpty else { return windows }

        let membership = (root["membershipType"] as? String) ?? "this"
        if (root["isUnlimited"] as? Bool) == true {
            throw UsageProviderError.nothingMetered("Unlimited on the \(membership) plan — nothing to meter")
        }
        throw UsageProviderError.nothingMetered("The \(membership) plan has nothing for Cursor to meter yet")
    }

    /// A dollar-denominated bucket, used where a plan states a real ceiling.
    private static func spendWindow(
        _ any: Any?, id: String, label: String, resetsAt: Date?
    ) -> LimitWindow? {
        guard let bucket = any as? [String: Any],
              (bucket["enabled"] as? Bool) == true,
              let limit = (bucket["limit"] as? NSNumber)?.doubleValue, limit > 0,
              let used = (bucket["used"] as? NSNumber)?.doubleValue
        else { return nil }
        return LimitWindow(id: id, label: label, usedFraction: used / limit, resetsAt: resetsAt)
    }

    /// Cursor reports 0–100; the rest of the app works in 0–1.
    private static func percent(_ any: Any?) -> Double? {
        guard let number = any as? NSNumber else { return nil }
        return number.doubleValue / 100
    }

    private static func date(_ any: Any?) -> Date? {
        guard let text = any as? String else { return nil }
        let withFraction = ISO8601DateFormatter()
        withFraction.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = withFraction.date(from: text) { return date }
        let plain = ISO8601DateFormatter()
        plain.formatOptions = [.withInternetDateTime]
        return plain.date(from: text)
    }
}

File: windows/QuotaArc/Providers/CursorUsage.cs
using System.Globalization;
using System.Text.Json;
using QuotaArc.Model;

namespace QuotaArc.Providers;

internal static class CursorUsage
{
    public static List<LimitWindow> WindowsFromJson(string json)
    {
        using var doc = JsonDocument.Parse(json);
        var root = doc.RootElement;
        DateTime? resetsAt = null;
        if (root.TryGetProperty("billingCycleEnd", out var end) &&
            DateTimeOffset.TryParse(end.GetString(), CultureInfo.InvariantCulture,
                DateTimeStyles.RoundtripKind, out var dto))
            resetsAt = dto.LocalDateTime;

        var usage = root.TryGetProperty("individualUsage", out var u) ? u : default;
        var plan = usage.ValueKind == JsonValueKind.Object && usage.TryGetProperty("plan", out var p) ? p : default;

        var windows = new List<LimitWindow>();
        if (Percent(plan, "totalPercentUsed") is { } total)
            windows.Add(new LimitWindow("included", "Included usage", total, ResetsAt: resetsAt));
        if (Percent(plan, "apiPercentUsed") is { } api && api > 0)
            windows.Add(new LimitWindow("api", "API usage", api, ResetsAt: resetsAt));
        if (SpendWindow(usage, "onDemand", "on_demand", "On demand", resetsAt) is { } od)
            windows.Add(od);

        if (windows.Count > 0) return windows;

        var membership = root.TryGetProperty("membershipType", out var m) ? m.GetString() ?? "this" : "this";
        if (root.TryGetProperty("isUnlimited", out var unl) && unl.ValueKind == JsonValueKind.True)
            throw UsageProviderException.NothingMetered($"Unlimited on the {membership} plan — nothing to meter");
        throw UsageProviderException.NothingMetered($"The {membership} plan has nothing for Cursor to meter yet");
    }

    private static LimitWindow? SpendWindow(JsonElement usage, string key, string id, string label, DateTime? resetsAt)
    {
        if (usage.ValueKind != JsonValueKind.Object || !usage.TryGetProperty(key, out var bucket))
            return null;
        if (!bucket.TryGetProperty("enabled", out var en) || en.ValueKind != JsonValueKind.True)
            return null;
        if (!TryDouble(bucket, "limit", out var limit) || limit <= 0) return null;
        if (!TryDouble(bucket, "used", out var used)) return null;
        return new LimitWindow(id, label, used / limit, ResetsAt: resetsAt);
    }

    private static double? Percent(JsonElement plan, string name)
    {
        if (plan.ValueKind != JsonValueKind.Object) return null;
        return TryDouble(plan, name, out var n) ? n / 100 : null;
    }

    private static bool TryDouble(JsonElement el, string name, out double value)
    {
        value = 0;
        if (!el.TryGetProperty(name, out var n) || n.ValueKind != JsonValueKind.Number) return false;
        value = n.GetDouble();
        return true;
    }
}
